geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/StrategyDecider.scala

```diff
@@ -0,0 +1,145 @@
+/***********************************************************************
+ * Copyright (c) 2013-2017 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.index.planning
+
+import com.typesafe.scalalogging.LazyLogging
+import org.locationtech.geomesa.index.api._
+import org.locationtech.geomesa.index.geotools.GeoMesaDataStore
+import org.locationtech.geomesa.index.planning.QueryPlanner.CostEvaluation
+import org.locationtech.geomesa.index.planning.QueryPlanner.CostEvaluation.CostEvaluation
+import org.locationtech.geomesa.index.stats.GeoMesaStats
+import org.locationtech.geomesa.index.utils.{ExplainNull, Explainer}
+import org.locationtech.geomesa.utils.conf.GeoMesaSystemProperties.SystemProperty
+import org.locationtech.geomesa.utils.index.IndexMode
+import org.locationtech.geomesa.utils.stats.{MethodProfiling, Timing, TimingsImpl}
+import org.opengis.feature.simple.SimpleFeatureType
+import org.opengis.filter.Filter
+
+/**
+  * Selects from available strategies for answering a query
+  */
+trait StrategyDecider {
+
+  /**
+    * Select from available filter plans
+    *
+    * @param sft simple feature type being queried
+    * @param stats handle to stats service, if available
+    * @param options filter plans that can satisfy the query
+    * @param transform requested query transform, if any
+    * @param explain explain logging
+    * @return filter plan to execute
+    */
+  def selectFilterPlan[DS <: GeoMesaDataStore[DS, F, W], F <: WrappedFeature, W]
+      (sft: SimpleFeatureType,
+       stats: Option[GeoMesaStats],
+       options: Seq[FilterPlan[DS, F, W]],
+       transform: Option[SimpleFeatureType],
+       explain: Explainer): FilterPlan[DS, F, W]
+}
+
+class CostBasedStrategyDecider extends StrategyDecider with MethodProfiling {
+  override def selectFilterPlan[DS <: GeoMesaDataStore[DS, F, W], F <: WrappedFeature, W]
+      (sft: SimpleFeatureType,
+       stats: Option[GeoMesaStats],
+       options: Seq[FilterPlan[DS, F, W]],
+       transform: Option[SimpleFeatureType],
+       explain: Explainer): FilterPlan[DS, F, W] = {
+    val costs = options.map { option =>
+      implicit val timing = new Timing()
+      val optionCosts = profile(option.strategies.map(f => f.index.getCost(sft, stats, f, transform)))
+      (option, optionCosts.sum, timing.time)
+    }.sortBy(_._2)
+    explain(s"Costs: ${costs.map(c => s"${c._1} (Cost ${c._2} in ${c._3}ms)").mkString("; ")}")
+    costs.head._1
+  }
+}
+
+object StrategyDecider extends MethodProfiling with LazyLogging {
+
+  private val decider: StrategyDecider = SystemProperty("geomesa.strategy.decider").option match {
```

Comment:
[thought] should we make this a ```def``` instead of ```val``` in order to allow for dynamically reloading it? May be interesting if you want to change it on the fly? could always be later i don't particularly have a concrete use case other than my thought that i would want to possibly change it without restarting a server but currently there's no way to do that.